pkg/ring/model.go

```diff
@@ -442,3 +503,229 @@ func (d *Desc) RemoveTombstones(limit time.Time) {
 		}
 	}
 }
+
+// findTokenIndex searches the ring for a token. Assumes that the ring is already denormalised.
+func (d *Desc) findTokenIndex(token uint32) (int, bool) {
+	i := sort.Search(len(d.Tokens), func(x int) bool {
+		return d.Tokens[x].Token >= token
+	})
+	if i >= len(d.Tokens) {
+		i = 0
+	}
+	return i, len(d.Tokens) > i && d.Tokens[i].Token == token
+}
+
+// HealthCheckFunc is a function that validates whether a given token in the
+// ring is healthy and can be used.
+type HealthCheckFunc func(TokenDesc) bool
+
+// HealthChecker generates a health check function that wraps around
+// IngesterDesc.IsHealthyState, validating both the token state and the
+// last heartbeat of the ingester.
+func (d *Desc) HealthChecker(op Operation, heartbeatTimeout time.Duration) HealthCheckFunc {
+	return func(t TokenDesc) bool {
+		ing := d.Ingesters[t.Ingester]
+		return ing.IsHealthy(op, heartbeatTimeout)
+	}
+}
+
+// Neighbor moves around the ring to find the nth neighboring ingester of start.
+// The healthy function determines which tokens are considered as neighbors.
+//
+// Only one token per ingester is considred as a neighbor. If excludeStart is
+// true, then the ingester owning opts.Start can not be considered as a neighbor.
+//
+// If opts.Offset has a positive value, Neighbor searches the ring clockwise.
+// Otherwise, if offset has a negative value, Neighbor searches the ring counter-clockwise.
+// An offset of 0 will return the starting token.
+func (d *Desc) Neighbor(start uint32, offset int, excludeStart bool, healthy HealthCheckFunc) (TokenDesc, error) {
+	idx, ok := d.findTokenIndex(start)
+	if !ok {
+		return TokenDesc{}, fmt.Errorf("could not find token %d in ring", start)
+	}
+	if offset == 0 {
+		return d.Tokens[idx], nil
+	}
+
+	distinct := map[string]bool{}
+	if excludeStart {
+		distinct[d.Tokens[idx].Ingester] = true
+	}
+
+	numNeighbors := offset
+	neighborCount := 0
+
+	direction := 1
+
+	if offset < 0 {
+		direction = -1
+		numNeighbors = -numNeighbors
+	}
+
+	it := newTokenIterator(d, direction, idx, healthy)
+	for it.HasNext() {
+		successor := it.Next()
+		if distinct[successor.Ingester] {
+			continue
+		}
+
+		neighborCount++
+		if neighborCount == numNeighbors {
+			return successor, nil
+		}
+
+		distinct[successor.Ingester] = true
+	}
+
+	return TokenDesc{}, fmt.Errorf("could not find neighbor #%d for token %d", offset, start)
+}
+
+// Predecessors finds all tokens of which the start is the offset-th neighbor to.
```

Comment:
So here's what's happening with the error messages: 

1. We want to know the set of tokens where 518 is the 5th successor (based on `5:`) there. 
2. We work backwards from 518, starting at 505. 
3. We get the 5th successor for 505 and check if it's 518. 
4. There is no 5th successor for 505. There are only 5 ingesters in the ring total, so there will be no more than 4 successors. 

However, something seems fishy with offsets 0-4 returning empty slices. Offset 1 should return `[{295 D}]`. How are you generating the ring?

Reply:
This is the test code I use: https://gist.github.com/pstibrany/4131a178dfa03866901dc1faf4f38568

(I had to modify some methods to be callable like this, see full.patch in same gist)

Reply:
Took me a while to catch it, but the output is correct here. Because token `505` is right before `518`, `505` will always be the successor chosen by all the previous tokens. This is actually a case we _do_ want to catch, because it means that `518` has nothing to transfer for the non-zero offsets. 

For zero offsets (i.e., `518` itself), we'd still want to transfer stuff in the `505`-`518` range (which E owns), so this makes sense too.

Reply:
Thanks for explanation, it makes sense to me now.